OpenRA.Mods.Common/AI/HackyAI.cs

```diff
@@ -1005,7 +1010,10 @@ void BuildUnit(string category, bool buildRandom)
 				ChooseRandomUnitToBuild(queue) :
 				ChooseUnitToBuild(queue);
 
-			if (unit != null && Info.UnitsToBuild.Any(u => u.Key == unit.Name))
+			if (unit != null && Info.UnitsToBuild.Any(u => u.Key == unit.Name
+				&& (!Info.UnitLimits.ContainsKey(unit.Name)
+					|| World.Actors.Where(a => a.Owner == Player).Count(a => a.Info.Name == unit.Name) < Info.UnitLimits[unit.Name])))
```

Comment:
Why is this bad for perf? The LINQ is only checked if the unit exists in `UnitLimits`. In most cases, it just gets expanded with a dictionary lookup.

Caching all actors in HackyAI isn't hard per se, but that would bring in a bigger refactor to update all places where counting units are used.